File: Tele-Bot-Api/src/services/ConnectionManager.ts
import { ApiError } from '@/errors';
import { useTry, useTryAsync } from 'no-try';
import { Api, TelegramClient } from 'telegram';

export class ConnectionManager {
  constructor(private client: TelegramClient) {}

  async isConnectionReady() {
    const [err, data] =
      (await useTryAsync(() => this.client.isUserAuthorized())) ||
      this.client.connected;

    if (err) {
      throw new ApiError('Connection check failed.', err.message);
    }
    return data;
  }

  async connect() {
    const [err] = await useTryAsync(() => this.client.connect());

    if (err) {
      throw new ApiError('Connection failed.', err.message);
    }
  }

  async disconnect() {
    const [err] = await useTryAsync(() => this.client.disconnect());

    if (err) {
      throw new ApiError('Disconnection failed.', err.message);
    }
  }

  async destroy() {
    const [err] = await useTryAsync(() => this.client.destroy());

    if (err) {
      throw new ApiError('Disconnection failed.', err.message);
    }
  }

  async start(phoneNumber: string, phoneCode: string) {
    const [err] = await useTryAsync(() =>
      this.client.start({
        phoneNumber,
        phoneCode: async () => phoneCode,
        onError: console.error,
      }),
    );

    if (err) {
      throw new ApiError('Starting failed.', err.message, {
        phoneNumber,
      });
    }

    console.log('You are now connected.');
  }

  async signIn(phoneNumber: string, phoneCode: string, codeHash: string) {
    const [err] = await useTryAsync(() =>
      this.client.invoke(
        new Api.auth.SignIn({
          phoneNumber: phoneNumber,
          phoneCode: phoneCode,
          phoneCodeHash: codeHash,
        }),
      ),
    );

    if (err) {
      throw new ApiError('Starting failed.', err.message, {
        phoneNumber,
      });
    }

    console.log('You are now connected.');
  }

  async sendCode(phoneNumber: string) {
    const [err, data] = await useTryAsync(() =>
      this.client.sendCode(
        { apiHash: this.client.apiHash, apiId: this.client.apiId },
        phoneNumber,
      ),
    );

    if (err) {
      throw new ApiError('Sending code failed.', err.message, {
        phoneNumber,
      });
    }

    return data;
  }

  saveSessionString() {
    const [err, data] = useTry(() => this.client.session.save());
    if (err) {
      throw new ApiError('Saving session failed.', err.message);
    }
    return data as unknown as string;
  }
}

File: Tele-Bot-Api-Py/app/services/connection_manager.py
from fastapi import HTTPException
from telethon import TelegramClient
from telethon.tl.functions.account import UpdateProfileRequest


class ConnectionManager:
    def __init__(self, client: TelegramClient):
        self.client = client

    def is_connected(self):
        return self.client.is_connected()

    async def is_banned(self):
        try:
            print(await self.client.get_me())
            return False
        except Exception as e:
            print(e)
            return True

    async def connect(self):
        try:
            await self.client.connect()
        except Exception as e:
            print(e)
            raise Exception(f"Error connecting to Telegram: {str(e)}")

    async def disconnect(self):
        try:
            await self.client.disconnect()
        except Exception as e:
            raise HTTPException(status_code=500, detail={
                "message": "Error disconnecting from Telegram",
                "error": str(e)
            })

    async def start(self, phone_number, phone_code):
        try:
            await self.client.start(phone=phone_number)
        except Exception as e:
            raise HTTPException(status_code=500, detail={
                "message": "Error starting Telegram client",
                "error": str(e)
            })

    async def send_code(self, phone_number):
        try:
            return await self.client.send_code_request(phone_number)
        except Exception as e:
            raise HTTPException(status_code=500, detail={
                "message": "Error sending code to phone number",
                "error": str(e)
            })

    async def sign_in(self, phone_number, phone_code, phone_code_hash=None):
        try:
            return await self.client.sign_in(phone=phone_number, code=phone_code, phone_code_hash=phone_code_hash)
        except Exception as e:
            raise HTTPException(status_code=500, detail={
                "message": "Error signing in to Telegram",
                "error": str(e)
            })

    async def get_entity(self, entity):
        try:
            return await self.client.get_entity(entity)
        except Exception as e:
            raise HTTPException(status_code=500, detail={
                "message": "Error getting entity",
                "error": str(e)
            })

    async def rename_bot(self, first_name, last_name=""):
        try:
            await self.client(UpdateProfileRequest(
                first_name=first_name,
                last_name=last_name
            ))
        except Exception as e:
            raise HTTPException(status_code=500, detail={
                "message": "Error renaming bot",
                "error": str(e)
            })

    def save_session_string(self):
        return self.client.session.save()
